src/server/journal/serializer.cc

```diff
@@ -0,0 +1,222 @@
+      return Write(entry.dbid);
+    case journal::Op::COMMAND:
+      RETURN_ON_ERR(Write(entry.txid));
+      return std::visit([this](const auto& payload) { return Write(payload); }, entry.payload);
+    default:
+      break;
+  };
+  return std::error_code{};
+}
+
+JournalReader::JournalReader(io::Source* source, DbIndex dbid)
+    : source_{source}, buf_{}, dbid_{dbid} {
+}
+
+io::Result<uint8_t> JournalReader::ReadU8() {
+  uint8_t buf[1];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 1)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
+  return buf[0];
+}
+
+io::Result<uint16_t> JournalReader::ReadU16() {
+  uint8_t buf[2];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 2)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
+  return absl::big_endian::Load16(buf);
+}
+
+io::Result<uint64_t> JournalReader::ReadU64() {
+  uint8_t buf[8];
+  size_t read;
+  SET_OR_UNEXPECT(source_->Read(io::MutableBytes{buf}), read);
+  if (read < 8)  // TODO: Custom errc namespace? Generic opcodes?
+    return make_unexpected(std::make_error_code(std::errc::result_out_of_range));
+  return absl::big_endian::Load64(buf);
+}
+
+io::Result<size_t> JournalReader::ReadString() {
+  size_t size = 0;
+  SET_OR_UNEXPECT(ReadU64(), size);
+
+  buf_.Reserve(buf_.InputLen() + size);  // TODO: There should be a more intuitive function
+
+  auto dest = buf_.AppendBuffer().first(size);
+  auto res = source_->Read(dest);
+  if (!res)
+    return res;
+
+  buf_.CommitWrite(*res);
+  if (*res != size)
+    return make_unexpected(std::make_error_code(std::errc::message_size));
+
+  return size;
+}
+
+std::error_code JournalReader::Read(CmdArgVec* vec) {
+  buf_.ConsumeInput(buf_.InputBuffer().size());
+
+  size_t size = 0;
+  SET_OR_RETURN(ReadU64(), size);
+
+  vec->resize(size);
+  for (auto& span : *vec) {
+    size_t len;
+    SET_OR_RETURN(ReadString(), len);
+    span = MutableSlice{nullptr, len};
+  }
+
+  size_t offset = 0;
+  for (auto& span : *vec) {
+    size_t len = span.size();
+    auto ptr = buf_.InputBuffer().subspan(offset).data();
+    span = MutableSlice{reinterpret_cast<char*>(ptr), len};
+    offset += len;
+  }
+
+  return std::error_code{};
+}
+
+io::Result<journal::Entry> JournalReader::ReadEntry() {
+  uint8_t opcode;
+  SET_OR_UNEXPECT(ReadU8(), opcode);
+
+  journal::Entry entry{static_cast<journal::Op>(opcode), dbid_};
+
+  switch (entry.opcode) {
+    case journal::Op::COMMAND:
+      SET_OR_UNEXPECT(ReadU64(), entry.txid);
+      entry.owned_payload = CmdArgVec{};
+      if (auto ec = Read(&*entry.owned_payload); ec)
+        return make_unexpected(ec);
+      break;
+    case journal::Op::SELECT:
+      SET_OR_UNEXPECT(ReadU16(), dbid_);
+      return ReadEntry();
```

Comment:
I decided to do this inside the reader instead of the executor so that the number of read entries is consistent

Reply:
This is because I write the number of entries to be parsed when I do a journal blob inside rdb

alternatively we can write a journal stop opcode